File: invesdwin-scripting-parent/invesdwin-scripting-julia/invesdwin-scripting-julia-runtime-contract/src/test/java/de/invesdwin/scripting/julia/runtime/contract/callback/ParametersAndReturnsTestString.jl
println("getString")
if isdefined(Main, :getString) && !isnothing(getString)
	error("getString already defined!")
end
getString = callback("getString")
println(typeof(getString))
println(getString)
if typeof(getString) != String
	error("getString not String!")
end
callback("setString",getString)

println("getStringWithNull")
if isdefined(Main, :getStringWithNull) && !isnothing(getStringWithNull)
	error("getStringWithNull already defined!")
end
getStringWithNull = callback("getStringWithNull")
println(typeof(getStringWithNull))
println(getStringWithNull)
if typeof(getStringWithNull) != Nothing
	error("getStringWithNull not Nothing!")
end
if !isnothing(getStringWithNull)
	error("getStringWithNull not nothing!")
end
callback("setStringWithNull",getStringWithNull)

println("getStringVector")
if isdefined(Main, :getStringVector) && !isnothing(getStringVector)
	error("getStringVector already defined!")
end
getStringVector = callback("getStringVector")
println(eltype(getStringVector))
println(getStringVector)
if eltype(getStringVector) != String
	error("getStringVector not String!")
end
callback("setStringVector",getStringVector)


println("getStringVectorWithNull")
if isdefined(Main, :getStringVectorWithNull) && !isnothing(getStringVectorWithNull)
	error("getStringVectorWithNull already defined!")
end
getStringVectorWithNull = callback("getStringVectorWithNull")
println(eltype(getStringVectorWithNull))
println(getStringVectorWithNull)
if eltype(getStringVectorWithNull) != String
	error("getStringVectorWithNull not String!")
end
if !isempty(getStringVectorWithNull[2])
	error("getStringVectorWithNull[2] not empty!")
end
callback("setStringVectorWithNull",getStringVectorWithNull)

println("getStringVectorAsList")
if isdefined(Main, :getStringVectorAsList) && !isnothing(getStringVectorAsList)
	error("getStringVectorAsList already defined!")
end
getStringVectorAsList = callback("getStringVectorAsList")
println(eltype(getStringVectorAsList))
println(getStringVectorAsList)
if eltype(getStringVectorAsList) != String
	error("getStringVectorAsList not String!")
end
callback("setStringVectorAsList",getStringVectorAsList)

println("getStringVectorAsListWithNull")
if isdefined(Main, :getStringVectorAsListWithNull) && !isnothing(getStringVectorAsListWithNull)
	error("getStringVectorAsListWithNull already defined!")
end
getStringVectorAsListWithNull = callback("getStringVectorAsListWithNull")
println(eltype(getStringVectorAsListWithNull))
println(getStringVectorAsListWithNull)
if eltype(getStringVectorAsListWithNull) != String
	error("getStringVectorAsListWithNull not String!")
end
if !isempty(getStringVectorAsListWithNull[2])
	error("getStringVectorAsListWithNull[2] not empty!")
end
callback("setStringVectorAsListWithNull",getStringVectorAsListWithNull)

println("getStringMatrix")
if isdefined(Main, :getStringMatrix) && !isnothing(getStringMatrix)
	error("getStringMatrix already defined!")
end
getStringMatrix = callback("getStringMatrix")
println(eltype(getStringMatrix))
println(getStringMatrix)
if eltype(getStringMatrix) != String
	error("getStringMatrix not String!")
end
callback("setStringMatrix",getStringMatrix)


println("getStringMatrixWithNull")
if isdefined(Main, :getStringMatrixWithNull) && !isnothing(getStringMatrixWithNull)
	error("getStringMatrixWithNull already defined!")
end
getStringMatrixWithNull = callback("getStringMatrixWithNull")
println(eltype(getStringMatrixWithNull))
println(getStringMatrixWithNull)
if eltype(getStringMatrixWithNull) != String
	error("getStringMatrixWithNull not String!")
end
if !isempty(getStringMatrixWithNull[1,1])
	error("getStringMatrixWithNull[1,1] not empty!")
end
if !isempty(getStringMatrixWithNull[2,2])
	error("getStringMatrixWithNull[2,2] not empty!")
end
if !isempty(getStringMatrixWithNull[3,3])
	error("getStringMatrixWithNull[3,3] not empty!")
end
callback("setStringMatrixWithNull",getStringMatrixWithNull)

println("getStringMatrixAsList")
if isdefined(Main, :getStringMatrixAsList) && !isnothing(getStringMatrixAsList)
	error("getStringMatrixAsList already defined!")
end
getStringMatrixAsList = callback("getStringMatrixAsList")
println(eltype(getStringMatrixAsList))
println(getStringMatrixAsList)
if eltype(getStringMatrixAsList) != String
	error("getStringMatrixAsList not String!")
end
callback("setStringMatrixAsList",getStringMatrixAsList)

println("getStringMatrixAsListWithNull")
if isdefined(Main, :getStringMatrixAsListWithNull) && !isnothing(getStringMatrixAsListWithNull)
	error("getStringMatrixAsListWithNull already defined!")
end
getStringMatrixAsListWithNull = callback("getStringMatrixAsListWithNull")
println(eltype(getStringMatrixAsListWithNull))
println(getStringMatrixAsListWithNull)
if eltype(getStringMatrixAsListWithNull) != String
	error("getStringMatrixAsListWithNull not String!")
end
if !isempty(getStringMatrixAsListWithNull[1,1])
	error("getStringMatrixAsListWithNull[1,1] not empty!")
end
if !isempty(getStringMatrixAsListWithNull[2,2])
	error("getStringMatrixAsListWithNull[2,2] not empty!")
end
if !isempty(getStringMatrixAsListWithNull[3,3])
	error("getStringMatrixAsListWithNull[3,3] not empty!")
end
callback("setStringMatrixAsListWithNull",getStringMatrixAsListWithNull)

File: invesdwin-scripting-parent/invesdwin-scripting-runtime-beanshell/src/test/java/de/invesdwin/scripting/runtime/beanshell/tests/callback/ParametersAndReturnsTestString.java
package de.invesdwin.scripting.runtime.beanshell.tests.callback;

import java.util.ArrayList;
import java.util.List;

import javax.annotation.concurrent.NotThreadSafe;

import org.springframework.core.io.ClassPathResource;

import de.invesdwin.scripting.IScriptTaskEngine;
import de.invesdwin.scripting.IScriptTaskInputs;
import de.invesdwin.scripting.IScriptTaskResults;
import de.invesdwin.scripting.callback.IScriptTaskCallback;
import de.invesdwin.scripting.callback.ReflectiveScriptTaskCallback;
import de.invesdwin.scripting.runtime.beanshell.AScriptTaskBeanshell;
import de.invesdwin.scripting.runtime.beanshell.IScriptTaskRunnerBeanshell;
import de.invesdwin.util.assertions.Assertions;
import de.invesdwin.util.collections.Arrays;
import de.invesdwin.util.lang.Objects;

@NotThreadSafe
public class ParametersAndReturnsTestString {

    private final IScriptTaskRunnerBeanshell runner;

    public ParametersAndReturnsTestString(final IScriptTaskRunnerBeanshell runner) {
        this.runner = runner;
    }

    public void testString() {
        new AScriptTaskBeanshell<Void>() {

            @Override
            public IScriptTaskCallback getCallback() {
                final ParametersAndReturnsTestStringCallback callback = new ParametersAndReturnsTestStringCallback();
                return new ReflectiveScriptTaskCallback(callback);
            }

            @Override
            public void populateInputs(final IScriptTaskInputs inputs) {}

            @Override
            public void executeScript(final IScriptTaskEngine engine) {
                engine.eval(new ClassPathResource(ParametersAndReturnsTestString.class.getSimpleName() + ".bsh",
                        ParametersAndReturnsTestString.class));
            }

            @Override
            public Void extractResults(final IScriptTaskResults results) {
                return null;
            }
        }.run(runner);
    }

    public static class ParametersAndReturnsTestStringCallback {

        //putString
        private final String putString;
        private final String putStringWithNull;

        //putStringVector
        private final String[] putStringVector;
        private final String[] putStringVectorWithNull;

        //putStringVectorAsList
        private final List<String> putStringVectorAsList;
        private final List<String> putStringVectorAsListWithNull;

        //putStringMatrix
        private final String[][] putStringMatrix;
        private final String[][] putStringMatrixWithNull;

        //putStringMatrixAsList
        private final List<List<String>> putStringMatrixAsList;
        private final List<List<String>> putStringMatrixAsListWithNull;

        public ParametersAndReturnsTestStringCallback() {
            //putString
            this.putString = "asdf";
            this.putStringWithNull = null;

            //putStringVector
            this.putStringVector = new String[3];
            for (int i = 0; i < putStringVector.length; i++) {
                putStringVector[i] = i + "-" + i;
            }
            this.putStringVectorWithNull = Objects.deepClone(putStringVector);
            putStringVectorWithNull[1] = null;

            //putStringVectorAsList
            this.putStringVectorAsList = Arrays.asList(putStringVector);
            this.putStringVectorAsListWithNull = Objects.deepClone(putStringVectorAsList);
            putStringVectorAsListWithNull.set(1, null);

            //putStringMatrix
            this.putStringMatrix = new String[4][];
            for (int i = 0; i < putStringMatrix.length; i++) {
                final String[] vector = new String[3];
                for (int j = 0; j < vector.length; j++) {
                    vector[j] = i + "" + j + "-" + i + "" + j;
                }
                putStringMatrix[i] = vector;
            }
            this.putStringMatrixWithNull = Objects.deepClone(putStringMatrix);
            for (int i = 0; i < putStringMatrixWithNull[0].length; i++) {
                putStringMatrixWithNull[i][i] = null;
            }

            //putStringMatrixAsList
            this.putStringMatrixAsList = new ArrayList<List<String>>(putStringMatrix.length);
            for (final String[] vector : putStringMatrix) {
                putStringMatrixAsList.add(Arrays.asList(vector));
            }
            this.putStringMatrixAsListWithNull = Objects.deepClone(putStringMatrixAsList);
            for (int i = 0; i < putStringMatrixAsListWithNull.get(0).size(); i++) {
                putStringMatrixAsListWithNull.get(i).set(i, null);
            }
        }

        public String getString() {
            return putString;
        }

        public void setString(final String putString) {
            Assertions.assertThat(this.putString).isEqualTo(putString);
        }

        public String getStringWithNull() {
            return putStringWithNull;
        }

        public void setStringWithNull(final String putStringWithNull) {
            Assertions.assertThat(this.putStringWithNull).isEqualTo(putStringWithNull);
        }

        public String[] getStringVector() {
            return putStringVector;
        }

        public void setStringVector(final String[] putStringVector) {
            Assertions.assertThat(this.putStringVector).isEqualTo(putStringVector);
        }

        public String[] getStringVectorWithNull() {
            return putStringVectorWithNull;
        }

        public void setStringVectorWithNull(final String[] putStringVectorWithNull) {
            Assertions.assertThat(this.putStringVectorWithNull).isEqualTo(putStringVectorWithNull);
        }

        public List<String> getStringVectorAsList() {
            return putStringVectorAsList;
        }

        public void setStringVectorAsList(final List<String> putStringVectorAsList) {
            Assertions.assertThat(this.putStringVectorAsList).isEqualTo(putStringVectorAsList);
        }

        public List<String> getStringVectorAsListWithNull() {
            return putStringVectorAsListWithNull;
        }

        public void setStringVectorAsListWithNull(final List<String> putStringVectorAsListWithNull) {
            Assertions.assertThat(this.putStringVectorAsListWithNull).isEqualTo(putStringVectorAsListWithNull);
        }

        public String[][] getStringMatrix() {
            return putStringMatrix;
        }

        public void setStringMatrix(final String[][] putStringMatrix) {
            Assertions.assertThat(this.putStringMatrix).isEqualTo(putStringMatrix);
        }

        public String[][] getStringMatrixWithNull() {
            return putStringMatrixWithNull;
        }

        public void setStringMatrixWithNull(final String[][] putStringMatrixWithNull) {
            Assertions.assertThat(this.putStringMatrixWithNull).isEqualTo(putStringMatrixWithNull);
        }

        public List<List<String>> getStringMatrixAsList() {
            return putStringMatrixAsList;
        }

        public void setStringMatrixAsList(final List<List<String>> putStringMatrixAsList) {
            Assertions.assertThat(this.putStringMatrixAsList).isEqualTo(putStringMatrixAsList);
        }

        public List<List<String>> getStringMatrixAsListWithNull() {
            return putStringMatrixAsListWithNull;
        }

        public void setStringMatrixAsListWithNull(final List<List<String>> putStringMatrixAsListWithNull) {
            Assertions.assertThat(this.putStringMatrixAsListWithNull).isEqualTo(putStringMatrixAsListWithNull);
        }

    }

}
